### PR TITLE
[block] mark dirtied mem when executing io_uring ops

### PR DESCRIPTION
**Pull request by serban300**
# Reason for This PR

[block] mark dirtied mem when executing io_uring ops

## Description of Changes

[block] mark dirtied mem when executing io_uring ops

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/devices/src/virtio/block/io/async_io.rs [hunk 1]
```diff
@@ -33,6 +33,35 @@ pub struct AsyncFileEngine<T> {
...
+            user_data,
+        }
+    }
+
+    fn new_with_addr(addr: GuestAddress, user_data: T) -> Self {
+        WrappedUserData {
+            addr: Some(addr),
+            user_data,
+        }
+    }
```
Comment (AlexandruCihodaru) [inline src/devices/src/virtio/block/io/async_io.rs:54]: should we merge these two?

```    
fn new(user_data: T, addr: Option<GuestAddress>) -> Self {
        WrappedUserData {
            addr: None,
            user_data,
        }
    }
```

File: src/devices/src/virtio/block/io/mod.rs [hunk 2]
```diff
@@ -157,10 +157,10 @@ pub mod tests {
     use super::*;
     use utils::kernel_version::KernelVersion;
     use utils::tempfile::TempFile;
-    use vm_memory::Bytes;
+    use vm_memory::{Bitmap, Bytes, GuestMemory};
 
     const FILE_LEN: u32 = 1024;
-    const MEM_LEN: usize = 4096;
```
Comment (AlexandruCihodaru) [inline src/devices/src/virtio/block/io/mod.rs:163]: please add a comment about where this value is taken from

**Review by AlexandruCihodaru — COMMENTED** [review]
Overall LGTM

File: src/devices/src/virtio/block/io/mod.rs [hunk 3]
```diff
@@ -198,32 +198,22 @@ pub mod tests {
...
-            if let FileEngine::Async(ref mut engine) = $engine {
-                engine.kick_submission_queue_and_wait(1).unwrap();
-                assert_eq!(engine.pop().unwrap().unwrap().result().unwrap(), $count);
-            }
-        };
+    fn assert_async_execution(engine: &mut FileEngine<()>, count: u32) {
+        if let FileEngine::Async(ref mut engine) = engine {
+            engine.kick_submission_queue_and_wait(1).unwrap();
+            assert_eq!(engine.pop().unwrap().unwrap().result().unwrap(), count);
+        }
```
Comment (luminitavoicu) [inline src/devices/src/virtio/block/io/mod.rs]: I don't really understand why this change is required. Could you please provide more detail?

File: src/devices/src/virtio/block/io/async_io.rs [hunk 4]
```diff
@@ -33,6 +33,35 @@ pub struct AsyncFileEngine<T> {
...
+
+    fn into_user_data(self, mem: &GuestMemoryMmap, count: u32) -> T {
```
Comment (luminitavoicu) [inline src/devices/src/virtio/block/io/async_io.rs:56]: I don't find this function's name explicit. I would use something like: `mark_user_data_dirty()` or something similar. I don't have a strong preference for this particular name, so I am open to other suggestions if you have ideas.

Comment (serban300) [reply src/devices/src/virtio/block/io/async_io.rs]: Makes sense. I renamed it to `mark_dirty_mem_and_unwrap()`. WDYT ?

**Review by AlexandruCihodaru — APPROVED** [review]
🚢

**Review by luminitavoicu — APPROVED** [review]
Looks good!
